File: 232ImplementQueueUsingStacks.py
from collections import deque
class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.Myqueue = deque()
        

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: void
        """
        self.Myqueue.append(x)

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        if len(self.Myqueue) > 0:
        	return self.Myqueue.popleft()
        else:
        	return None

    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        if len(self.Myqueue) > 0:
        	return self.Myqueue[0]
        else:
        	return None
        

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return len(self.Myqueue) == 0
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()



# use a list to implement it


    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.List=[]
        self.num=0


    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: void
        """
        self.List.append(x)
        self.num+=1

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        self.num-=1
        return self.List.pop(0)


    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        if self.empty():
            return NULL
        return self.List[0]

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return (self.num==0)

File: 232ImplementQueueUsingStacks.cpp
/*
Implement the following operations of a queue using stacks.

    push(x) -- Push element x to the back of queue.
    pop() -- Removes the element from in front of queue.
    peek() -- Get the front element.
    empty() -- Return whether the queue is empty.


Example:

MyQueue queue = new MyQueue();

queue.push(1);
queue.push(2);  
queue.peek();  // returns 1
queue.pop();   // returns 1
queue.empty(); // returns false

Notes:

    You must use only standard operations of a stack -- which means only push to top, peek/pop from top, size, and is empty operations are valid.
    Depending on your language, stack may not be supported natively. You may simulate a stack by using a list or deque (double-ended queue), as long as you use only standard operations of a stack.
    You may assume that all operations are valid (for example, no pop or peek operations will be called on an empty queue).

*/
// 
class MyQueue {
	stack<int>q1,helper;
public:
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        while(!q1.empty()){
            helper.push(q1.top());
            q1.pop();
        }
        q1.push(x);
        while(!helper.empty()){
            q1.push(helper.top());
            helper.pop();
        }
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
    	int res = q1.top();
        q1.pop();
        return res;
    }
    
    /** Get the front element. */
    int peek() {
       return q1.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return q1.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * bool param_4 = obj.empty();
 */
